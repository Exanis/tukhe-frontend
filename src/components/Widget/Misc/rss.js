import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { withStyles } from '@material-ui/core/styles';

import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import InfiniteScroll from 'react-infinite-scroller';

import RssForm from './rss.form';

export const classes = theme => ({
    link: {
        '& a': {
            textDecoration: 'none',
            color: theme.palette.primary.main,
            '&:visited': {
                color: theme.palette.primary.light
            },
            '&:hover': {
                textDecoration: 'underline'
            }
        }
    }
});

export class RawRss extends Component {
    state = {
        displayPages: 0
    };

    pseudoLoadData = (page) => this.setState({displayPages: page});

    render () {
        if (!this.props.data)
            return null;

        const items = this.props.data.slice(
            0, (this.state.displayPages + 1) * 10
        ).map(news => <ListItem button key={news.id}>
            <ListItemText
                className={this.props.classes.link}
                primary={<a href={news.href}>{news.title}</a>}
                secondary={this.props.intl.formatDate(news.date)}
                />
        </ListItem>);

        return <List className={this.props.containerClass}>
            <InfiniteScroll
                pageStart={0}
                loadMore={this.pseudoLoadData}
                hasMore={(this.state.displayPages + 1) * 10 < this.props.data.length}
                useWindow={false}
                >
            {items}
            </InfiniteScroll>
        </List>
    }

    componentDidMount() {
        this.props.startWidgetProcess({
            target: this.props.refreshWithoutAccount,
            interval: 300000,
            widget: this.props.uuid
        });
    }

    componentWillUnmount() {
        this.props.stopWidgetProcess({
            widget: this.props.uuid
        })
    }
}

const Rss = injectIntl(withStyles(classes)(RawRss));

Rss.propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string,
        href: PropTypes.string,
        title: PropTypes.string,
        date: PropTypes.string
    })),
    containerClass: PropTypes.string.isRequired,
    startWidgetProcess: PropTypes.func.isRequired,
    stopWidgetProcess: PropTypes.func.isRequired,
    refreshWithoutAccount: PropTypes.func.isRequired,
    uuid: PropTypes.string.isRequired
};

RawRss.propTypes = {
    ...Rss.propTypes,
    classes: PropTypes.object,
    intl: PropTypes.object
};

export default {
    Rss: Rss,
};

export const Forms = {
    Rss: RssForm
};
