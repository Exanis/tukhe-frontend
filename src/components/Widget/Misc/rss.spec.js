import React from 'react';
import { classes, RawRss } from './rss';
import { shallow } from 'enzyme';

const props = {
    data: [
        {
            id: 'rss1',
            href: 'http://example.org',
            title: 'RSS First element',
            date: '2019-01-01 01:01:01'
        }
    ],
    containerClass: 'aClass',
    startWidgetProcess: jest.fn(),
    stopWidgetProcess: jest.fn(),
    refreshWithoutAccount: jest.fn(),
    uuid: 'widget uuid',
    classes: {},
    intl: {
        formatDate: jest.fn()
    }
};

it('Should render correctly', () => {
    const wrapper = shallow(<RawRss {...props} />);

    expect(wrapper).toMatchSnapshot();
    expect(props.startWidgetProcess).toHaveBeenCalledWith({
        target: props.refreshWithoutAccount,
        interval: 300000,
        widget: 'widget uuid'
    });
    wrapper.unmount();
    expect(props.stopWidgetProcess).toHaveBeenCalledWith({
        widget: 'widget uuid'
    });
});

it('Should be able to change page', () => {
    const wrapper = shallow(<RawRss {...props} />);

    expect(wrapper.state('displayPages')).toEqual(0);
    wrapper.instance().pseudoLoadData(2);
    expect(wrapper.state('displayPages')).toEqual(2);
});

it('Should not crash without data', () => {
    const testProps = {
        ...props,
        data: null
    };
    const wrapper = shallow(<RawRss {...testProps} />);

    expect(wrapper).toMatchSnapshot();
});

it('Should keep theme consistant', () => {
    const theme = {
        palette: {
            primary: {
                main: '#000',
                light: '#aaa'
            }
        }
    };
    const style = classes(theme);

    expect(style).toMatchSnapshot();
});
