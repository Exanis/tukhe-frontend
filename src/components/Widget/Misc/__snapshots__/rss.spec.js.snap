// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should keep theme consistant 1`] = `
Object {
  "link": Object {
    "& a": Object {
      "&:hover": Object {
        "textDecoration": "underline",
      },
      "&:visited": Object {
        "color": "#aaa",
      },
      "color": "#000",
      "textDecoration": "none",
    },
  },
}
`;

exports[`Should not crash without data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RawRss
    classes={Object {}}
    containerClass="aClass"
    data={null}
    intl={
      Object {
        "formatDate": [MockFunction] {
          "calls": Array [
            Array [
              "2019-01-01 01:01:01",
            ],
            Array [
              "2019-01-01 01:01:01",
            ],
            Array [
              "2019-01-01 01:01:01",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    refreshWithoutAccount={[MockFunction]}
    startWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "interval": 300000,
              "target": [MockFunction],
              "widget": "widget uuid",
            },
          ],
          Array [
            Object {
              "interval": 300000,
              "target": [MockFunction],
              "widget": "widget uuid",
            },
          ],
          Array [
            Object {
              "interval": 300000,
              "target": [MockFunction],
              "widget": "widget uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    stopWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "widget": "widget uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    uuid="widget uuid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RawRss
    classes={Object {}}
    containerClass="aClass"
    data={
      Array [
        Object {
          "date": "2019-01-01 01:01:01",
          "href": "http://example.org",
          "id": "rss1",
          "title": "RSS First element",
        },
      ]
    }
    intl={
      Object {
        "formatDate": [MockFunction] {
          "calls": Array [
            Array [
              "2019-01-01 01:01:01",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    refreshWithoutAccount={[MockFunction]}
    startWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "interval": 300000,
              "target": [MockFunction],
              "widget": "widget uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    stopWidgetProcess={[MockFunction]}
    uuid="widget uuid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <InfiniteScroll
        element="div"
        getScrollParent={null}
        hasMore={false}
        initialLoad={true}
        isReverse={false}
        loadMore={[Function]}
        loader={null}
        pageStart={0}
        ref={null}
        threshold={250}
        useCapture={false}
        useWindow={false}
      >
        <WithStyles(ListItem)
          button={true}
        >
          <WithStyles(ListItemText)
            primary={
              <a
                href="http://example.org"
              >
                RSS First element
              </a>
            }
          />
        </WithStyles(ListItem)>
      </InfiniteScroll>,
      "className": "aClass",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(ListItem)
            button={true}
          >
            <WithStyles(ListItemText)
              primary={
                <a
                  href="http://example.org"
                >
                  RSS First element
                </a>
              }
            />
          </WithStyles(ListItem)>,
        ],
        "element": "div",
        "getScrollParent": null,
        "hasMore": false,
        "initialLoad": true,
        "isReverse": false,
        "loadMore": [Function],
        "loader": null,
        "pageStart": 0,
        "ref": null,
        "threshold": 250,
        "useCapture": false,
        "useWindow": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "rss1",
          "nodeType": "class",
          "props": Object {
            "button": true,
            "children": <WithStyles(ListItemText)
              primary={
                <a
                  href="http://example.org"
                >
                  RSS First element
                </a>
              }
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": undefined,
              "primary": <a
                href="http://example.org"
              >
                RSS First element
              </a>,
              "secondary": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <InfiniteScroll
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={true}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={false}
        >
          <WithStyles(ListItem)
            button={true}
          >
            <WithStyles(ListItemText)
              primary={
                <a
                  href="http://example.org"
                >
                  RSS First element
                </a>
              }
            />
          </WithStyles(ListItem)>
        </InfiniteScroll>,
        "className": "aClass",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(ListItem)
              button={true}
            >
              <WithStyles(ListItemText)
                primary={
                  <a
                    href="http://example.org"
                  >
                    RSS First element
                  </a>
                }
              />
            </WithStyles(ListItem)>,
          ],
          "element": "div",
          "getScrollParent": null,
          "hasMore": false,
          "initialLoad": true,
          "isReverse": false,
          "loadMore": [Function],
          "loader": null,
          "pageStart": 0,
          "ref": null,
          "threshold": 250,
          "useCapture": false,
          "useWindow": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "rss1",
            "nodeType": "class",
            "props": Object {
              "button": true,
              "children": <WithStyles(ListItemText)
                primary={
                  <a
                    href="http://example.org"
                  >
                    RSS First element
                  </a>
                }
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": undefined,
                "primary": <a
                  href="http://example.org"
                >
                  RSS First element
                </a>,
                "secondary": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
